### DISCUSSION TITLE
Parse and process a Ruby DSL for CRUD style updates[thoughts and suggestions]

### DISCUSSION THREAD

**Discussion by aaronlippold** [ast-grep/ast-grep#1567] [Q&A]
I work on a project that writes large numbers of inspec profiles - https://saf.mitre.org - and we would like to look at using ast-grep as the foundation for some of our update and maintenance tooling. InSpec is a ruby DSL and so I would like to see if I can setup a good pattern of rules and matchers to 'parse and process' the inspec profiles so that we can streamline the update and maitence of the codebase for these security profiles.

For example:

```ruby
# encoding: UTF-8

control "V-93005" do
  title "Windows Server 2019 Deny log on locally user right on domain
controllers must be configured to prevent unauthenticated access."
  desc  "Inappropriate granting of user rights can provide system,
administrative, and other high-level capabilities.

    The \"Deny log on locally\" user right defines accounts that are prevented
from logging on interactively.

    The Guests group must be assigned this right to prevent unauthenticated
access."
  desc  "rationale", ""
  desc 'check', "This applies to domain controllers. A separate version applies to other
systems.

    Verify the effective setting in Local Group Policy Editor.

    Run \"gpedit.msc\".

    Navigate to Local Computer Policy >> Computer Configuration >> Windows
Settings >> Security Settings >> Local Policies >> User Rights Assignment.

    If the following accounts or groups are not defined for the \"Deny log on
locally\" user right, this is a finding:

    - Guests Group

    For server core installations, run the following command:

    Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt

    Review the text file.

    If the following SID(s) are not defined for the
\"SeDenyInteractiveLogonRight\" user right, this is a finding:

    S-1-5-32-546 (Guests)"
  desc  'fix', "
    Configure the policy value for Computer Configuration >> Windows Settings
>> Security Settings >> Local Policies >> User Rights Assignment >> \"Deny log
on locally\" to include the following:

    - Guests Group"
  impact 0.5
  tag 'severity': nil
  tag 'gtitle': 'SRG-OS-000080-GPOS-00048'
  tag 'gid': 'V-93005'
  tag 'rid': 'SV-103093r1_rule'
  tag 'stig_id': 'WN19-DC-000400'
  tag 'fix_id': 'F-99251r1_fix'
  tag 'cci': ["CCI-000213"]
  tag 'nist': ["AC-3", "Rev_4"]

  domain_role = command('wmic computersystem get domainrole | Findstr /v DomainRole').stdout.strip
  os_type = command('Test-Path "$env:windir\explorer.exe"').stdout.strip

  if os_type == 'False'
     describe 'This system is a Server Core Installation, and a manual check will need to be performed with command Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt' do
      skip 'This system is a Server Core Installation, and a manual check will need to be performed with command Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt'
     end
  end
  if domain_role == '4' || domain_role == '5'
    describe security_policy do
     its('SeDenyInteractiveLogonRight') { should eq ['S-1-5-32-546'] }
    end
  else
    impact 0.0
    describe 'This system is not a domain controller, therefore this control is not applicable as it only applies to domain controllers' do
      skip 'This system is not a domain controller, therefore this control is not applicable as it only applies to domain controllers'
    end
  end
end
```

This is an example of one of the controls - and you can see the first have is the DSL metadata and the second half is the tests which is a mix of the DSL testing ( basically rspec with sugar ) and some ruby.

Do you have suggestion on how we might write a set of rules where we could parse, update and wright out updates.

**Comment by aaronlippold** [comment]
General context:

A control has a: 
- title, 
- description, 
- zero or more sub-descriptions, 
- an initial impact, 
- a set of tags and refs, and a few other elements from the DSL. 

These parts are driven from the "security benchmark" (aka human document) and then the control has the "ruby tests" driven by the DSL in the form of "rspec like describe statement or rspec expect statements." 

Given this is a Ruby DSL there may also be some "strait up ruby" as well. 

Generally anything that is not part of the DSL metadata we consider the "Control Code Block." As we "update an inspec profile to an updated security benchmark," we usually have to: 
1. Update the metadata and then
2. . "transplant the control code block" and then review via a diff. Basically seeing if the update "metadata guidance" updated our "tests." 

We would like to use ast-grep to help us break out the parts and sections of our profiles and controls to make this maintenance easier, faster, and smarter.

**Reply by HerringtonDarkholme** [reply]
Hi thanks for learning ast-grep. Would you like to give a minimal example of code rewrite?

**Comment by aaronlippold** [comment]
You can find out more of the details here: 

- https://mitre.github.io/saf-training/courses/beginner/07.html
- https://mitre.github.io/saf-training/courses/beginner/05.html

**Comment by aaronlippold** [comment]
Happy to once I’m back at my desk.

--------
Aaron Lippold

***@***.***

260-255-4779

twitter/aim/yahoo,etc.
'aaronlippold'

**Comment by aaronlippold** [comment]
Apologies for the delayed response.

The code example above is the Ruby DSL I need to parse. My goal is to develop patterns to parse and tokenize InSpec profiles and controls to enhance our 'delta' process. An overview of the DSL is documented here: [InSpec Profiles and Controls](https://docs.chef.io/inspec/profiles/controls).

This library would support our 'delta' function in the saf-cli, which helps us update our InSpec profiles. For more details, see the Wiki: [Delta (mitre/saf)](https://github.com/mitre/saf/wiki/Delta).

Our current work can be found here: [mitre/saf@main/src/commands/generate/delta.ts](https://github.com/mitre/saf/blob/main/src/commands/generate/delta.ts).

As you can see, we are essentially creating a parser ourselves, so utilizing tools like ast-grep and tree-sitter is very appealing to the team.

The primary challenge with parsing a mixed structure like this and automating its updates is separating the 'code body'—user-added code that implements the described security tests—from the DSL metadata that describes the intent of the test or requirement. This descriptive/metadata part of the 'control code' is generated from the security document, usually XML or CSV. To automate this, we need to distinguish between these two primary parts.

For example, the code body (human-added tests) from the above example is:

```ruby
domain_role = command('wmic computersystem get domainrole | Findstr /v DomainRole').stdout.strip
os_type = command('Test-Path "$env:windir\explorer.exe"').stdout.strip

if os_type == 'False'
  describe 'This system is a Server Core Installation, and a manual check will need to be performed with command Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt' do
    skip 'This system is a Server Core Installation, and a manual check will need to be performed with command Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt'
  end
end
if domain_role == '4' || domain_role == '5'
  describe security_policy do
    its('SeDenyInteractiveLogonRight') { should eq ['S-1-5-32-546'] }
  end
else
  impact 0.0
  describe 'This system is not a domain controller, therefore this control is not applicable as it only applies to domain controllers' do
    skip 'This system is not a domain controller, therefore this control is not applicable as it only applies to domain controllers'
  end
end
```

I would want to transfer the updated InSpec control body into the update controle generated from the XCCDF XML document of the security benchmark. However, we need to parse the original control and all its parts to 'find' the code body of the control. The challenge is that I can have random Ruby code anywhere in the file and still be valid.
 
For example, the `domain_role` and `os_type` variables could be at the top of the file or between the `desc` statements and still be fully valid. So, I need to parse and tokenize the whole control Ruby and DSL to know which is which and what needs to move. or in reality my control could look like this, up side down from our style guide but technically valid inspec and ruby.

```ruby
# encoding: UTF-8

domain_role = command('wmic computersystem get domainrole | Findstr /v DomainRole').stdout.strip
os_type = command('Test-Path "$env:windir\explorer.exe"').stdout.strip

control "V-93005" do
  if os_type == 'False'
    describe 'This system is a Server Core Installation, and a manual check will need to be performed with command Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt' do
      skip 'This system is a Server Core Installation, and a manual check will need to be performed with command Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt'
    end
  end

  if domain_role == '4' || domain_role == '5'
    describe security_policy do
      its('SeDenyInteractiveLogonRight') { should eq ['S-1-5-32-546'] }
    end
  else
    impact 0.0
    describe 'This system is not a domain controller, therefore this control is not applicable as it only applies to domain controllers' do
      skip 'This system is not a domain controller, therefore this control is not applicable as it only applies to domain controllers'
    end
  end

  title "Windows Server 2019 Deny log on locally user right on domain controllers must be configured to prevent unauthenticated access."
  desc  "Inappropriate granting of user rights can provide system, administrative, and other high-level capabilities.

    The \"Deny log on locally\" user right defines accounts that are prevented from logging on interactively.

    The Guests group must be assigned this right to prevent unauthenticated access."
  desc  "rationale", ""
  desc 'check', "This applies to domain controllers. A separate version applies to other systems.

    Verify the effective setting in Local Group Policy Editor.

    Run \"gpedit.msc\".

    Navigate to Local Computer Policy >> Computer Configuration >> Windows Settings >> Security Settings >> Local Policies >> User Rights Assignment.

    If the following accounts or groups are not defined for the \"Deny log on locally\" user right, this is a finding:

    - Guests Group

    For server core installations, run the following command:

    Secedit /Export /Areas User_Rights /cfg c:\\path\\filename.txt

    Review the text file.

    If the following SID(s) are not defined for the \"SeDenyInteractiveLogonRight\" user right, this is a finding:

    S-1-5-32-546 (Guests)"
  desc 'fix', "Configure the policy value for Computer Configuration >> Windows Settings >> Security Settings >> Local Policies >> User Rights Assignment >> \"Deny log on locally\" to include the following:

    - Guests Group"
  impact 0.5
  tag 'severity': nil
  tag 'gtitle': 'SRG-OS-000080-GPOS-00048'
  tag 'gid': 'V-93005'
  tag 'rid': 'SV-103093r1_rule'
  tag 'stig_id': 'WN19-DC-000400'
  tag 'fix_id': 'F-99251r1_fix'
  tag 'cci': ["CCI-000213"]
  tag 'nist': ["AC-3", "Rev_4"]
end
```

We have hundreds of profiles to maintain, so creating a parser for automating and creating a 'dependabot' kind of process to create PRs of updated profiles when new security benchmarks are released would streamline the process significantly.

Secondly, once we have this library, I could also create a 'linter' and 'formatter' for the profiles to rewrite them easily into our 'style guide' and update profiles even if there isn't a security benchmark update, to our 'current best style'.

I would love to chat with you directly about this as I think it is a really cool problem, and I believe your work could be a big help. Alternatively, I would appreciate any feedback if you think I am heading down the wrong path.

**Comment by aaronlippold** [comment]
Does that provide the background you needed?

**Reply by HerringtonDarkholme** [reply]
Hi, thanks for the explanation. I'm wondering what is the output of processing you are looking for? 
Having two parts:  'code body' and 'metadata'?

**Comment by HerringtonDarkholme** [comment]
Hi Aaron,

Thanks for reaching out! To address your needs, you can leverage ast-grep’s API for parsing and rewriting your InSpec profiles. 

Here’s a brief guide to get you started:

Setup: Install ast-grep and import the necessary modules.
Define Rules: Create rules to match and extract DSL metadata and control code blocks.
Parse Profiles: Use ast-grep to parse the profiles and apply the defined rules.
Update and Rewrite: Modify the parsed elements as needed and write the updated profiles back.
For detailed instructions and examples, check out the ast-grep API Usage Guide. https://ast-grep.github.io/guide/api-usage.html

Feel free to reach out if you have any questions or need further assistance!
